Upgrade to the latest version of mypy

The current version of mypy in `noxfile.py` is `0.910` which is more than 2 years old. We should update to the latest version.
https://github.com/googleapis/python-pubsub/blob/265f4106f499ec5d2d01a127ba192404c1836a28/noxfile.py#L35C1-L35C1

This change is needed to bump the mypy version
```
(py3816) partheniou@partheniou-vm-3:~/git/python-pubsub$ git diff
diff --git a/noxfile.py b/noxfile.py
index 3f9db8b..775e0be 100644
--- a/noxfile.py
+++ b/noxfile.py
@@ -32,7 +32,7 @@ BLACK_VERSION = "black[jupyter]==23.7.0"
 ISORT_VERSION = "isort==5.11.0"
 LINT_PATHS = ["docs", "google", "tests", "noxfile.py", "setup.py"]
 
-MYPY_VERSION = "mypy==0.910"
+MYPY_VERSION = "mypy==1.8.0"

```

In addition, the mypy failures would need to be addressed

```
(py3816) partheniou@partheniou-vm-3:~/git/python-pubsub$ nox -s mypy
nox > Running session mypy
nox > Creating virtual environment (virtualenv) using python3.8 in .nox/mypy
nox > python -m pip install -e '.[all]'
nox > python -m pip install mypy==1.8.0
nox > python -m pip install types-protobuf types-setuptools
nox > python -m pip install 'google-api-core[grpc]>=2.1.1' 'google-cloud-core>=2.2.0'
nox > mypy -p google.cloud
google/cloud/pubsub_v1/subscriber/_protocol/helper_threads.py:35: error: Incompatible default for argument "max_items" (default has type "None", argument has type "int")  [assignment]
google/cloud/pubsub_v1/subscriber/_protocol/helper_threads.py:35: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
google/cloud/pubsub_v1/subscriber/_protocol/helper_threads.py:35: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
google/cloud/pubsub_v1/subscriber/scheduler.py:165: error: Value of type "Iterable[Any]" is not indexable  [index]
google/cloud/pubsub_v1/publisher/_sequencer/unordered_sequencer.py:110: error: Argument "batch_done_callback" has incompatible type "None"; expected "Callable[[bool], Any]"  [arg-type]
google/cloud/pubsub_v1/subscriber/futures.py:107: error: Argument 1 of "result" is incompatible with supertype "Future"; supertype defines the argument type as "Optional[float]"  [override]
google/cloud/pubsub_v1/subscriber/futures.py:107: note: This violates the Liskov substitution principle
google/cloud/pubsub_v1/subscriber/futures.py:107: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
google/cloud/pubsub_v1/subscriber/futures.py:107: error: Incompatible default for argument "timeout" (default has type "None", argument has type "Union[int, float]")  [assignment]
google/cloud/pubsub_v1/subscriber/futures.py:107: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
google/cloud/pubsub_v1/subscriber/futures.py:107: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
google/cloud/pubsub_v1/publisher/futures.py:48: error: Argument 1 of "result" is incompatible with supertype "Future"; supertype defines the argument type as "Optional[float]"  [override]
google/cloud/pubsub_v1/publisher/futures.py:48: note: This violates the Liskov substitution principle
google/cloud/pubsub_v1/publisher/futures.py:48: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
google/cloud/pubsub_v1/publisher/futures.py:48: error: Incompatible default for argument "timeout" (default has type "None", argument has type "Union[int, float]")  [assignment]
google/cloud/pubsub_v1/publisher/futures.py:48: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
google/cloud/pubsub_v1/publisher/futures.py:48: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
google/cloud/pubsub_v1/publisher/_batch/thread.py:93: error: Incompatible default for argument "batch_done_callback" (default has type "None", argument has type "Callable[[bool], Any]")  [assignment]
google/cloud/pubsub_v1/publisher/_batch/thread.py:93: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
google/cloud/pubsub_v1/publisher/_batch/thread.py:93: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py:48: error: Module "google.rpc" has no attribute "status_pb2"  [attr-defined]
google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py:270: error: Incompatible default for argument "scheduler" (default has type "None", argument has type "ThreadScheduler")  [assignment]
google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py:270: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py:270: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
google/cloud/pubsub_v1/subscriber/client.py:254: error: Argument "scheduler" to "StreamingPullManager" has incompatible type "Optional[ThreadScheduler]"; expected "ThreadScheduler"  [arg-type]
Found 11 errors in 8 files (checked 34 source files)
nox > Command mypy -p google.cloud failed with exit code 1
nox > Session mypy failed.
```

--- a/google/cloud/pubsub_v1/publisher/_batch/thread.py
+++ b/google/cloud/pubsub_v1/publisher/_batch/thread.py
@@ -90,7 +90,7 @@ class Batch(base.Batch):
         client: "PublisherClient",
         topic: str,
         settings: "types.BatchSettings",
-        batch_done_callback: Callable[[bool], Any] = None,
+        batch_done_callback: Optional[Callable[[bool], Any]] = None,
         commit_when_full: bool = True,
         commit_retry: "OptionalRetry" = gapic_v1.method.DEFAULT,
         commit_timeout: "types.OptionalTimeout" = gapic_v1.method.DEFAULT,

--- a/google/cloud/pubsub_v1/publisher/futures.py
+++ b/google/cloud/pubsub_v1/publisher/futures.py
@@ -45,7 +45,7 @@ class Future(futures.Future):
         """
         return False
 
-    def result(self, timeout: Union[int, float] = None) -> str:
+    def result(self, timeout: Union[int, float, None] = None) -> str:
         """Return the message ID or raise an exception.
 
         This blocks until the message has been published successfully and

--- a/google/cloud/pubsub_v1/subscriber/_protocol/helper_threads.py
+++ b/google/cloud/pubsub_v1/subscriber/_protocol/helper_threads.py
@@ -15,7 +15,7 @@
 import logging
 import queue
 import time
-from typing import Any, Callable, List, Sequence
+from typing import Any, Callable, List, Sequence, Optional
 import uuid
 
 
@@ -32,7 +32,7 @@ STOP = uuid.uuid4()
 
 
 def _get_many(
-    queue_: queue.Queue, max_items: int = None, max_latency: float = 0
+    queue_: queue.Queue, max_items: Optional[int] = None, max_latency: float = 0
 ) -> List[Any]:
     """Get multiple items from a Queue.
 

--- a/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py
+++ b/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py
@@ -267,7 +267,7 @@ class StreamingPullManager(object):
         client: "subscriber.Client",
         subscription: str,
         flow_control: types.FlowControl = types.FlowControl(),
-        scheduler: ThreadScheduler = None,
+        scheduler: Optional[ThreadScheduler] = None,
         use_legacy_flow_control: bool = False,
         await_callbacks_on_shutdown: bool = False,
     ):

--- a/google/cloud/pubsub_v1/subscriber/futures.py
+++ b/google/cloud/pubsub_v1/subscriber/futures.py
@@ -104,7 +104,7 @@ class Future(futures.Future):
         """
         return False
 
-    def result(self, timeout: Union[int, float] = None) -> AcknowledgeStatus:
+    def result(self, timeout: Union[int, float, None] = None) -> AcknowledgeStatus:
         """Return a success code or raise an exception.
 
         This blocks until the operation completes successfully and

--- a/google/cloud/pubsub_v1/subscriber/scheduler.py
+++ b/google/cloud/pubsub_v1/subscriber/scheduler.py
@@ -162,7 +162,7 @@ class ThreadScheduler(Scheduler):
                 work_item = self._executor._work_queue.get(block=False)
                 if work_item is None:  # Exceutor in shutdown mode.
                     continue
-                dropped_messages.append(work_item.args[0])
+                dropped_messages.append(work_item.args[0])  # type: ignore[index]
         except queue.Empty:
             pass
 

--- a/noxfile.py
+++ b/noxfile.py
@@ -32,10 +32,9 @@ BLACK_VERSION = "black[jupyter]==23.7.0"
 ISORT_VERSION = "isort==5.11.0"
 LINT_PATHS = ["docs", "google", "tests", "noxfile.py", "setup.py"]
 
-MYPY_VERSION = "mypy==0.910"
+MYPY_VERSION = "mypy==1.10.0"
 
 DEFAULT_PYTHON_VERSION = "3.8"
-
 UNIT_TEST_PYTHON_VERSIONS: List[str] = ["3.7", "3.8", "3.9", "3.10", "3.11", "3.12"]
 UNIT_TEST_STANDARD_DEPENDENCIES = [
     "mock",

--- a/noxfile.py
+++ b/noxfile.py
@@ -35,6 +35,7 @@ LINT_PATHS = ["docs", "google", "tests", "noxfile.py", "setup.py"]
 MYPY_VERSION = "mypy==1.10.0"
 
 DEFAULT_PYTHON_VERSION = "3.8"
+
 UNIT_TEST_PYTHON_VERSIONS: List[str] = ["3.7", "3.8", "3.9", "3.10", "3.11", "3.12"]
 UNIT_TEST_STANDARD_DEPENDENCIES = [
     "mock",

--- a/owlbot.py
+++ b/owlbot.py
@@ -348,7 +348,7 @@ s.move(templated_files, excludes=[".coveragerc", ".github/release-please.yml", "
 s.replace(
     "noxfile.py",
     r"LINT_PATHS = \[.*?\]",
-    '\g<0>\n\nMYPY_VERSION = "mypy==0.910"',
+    '\g<0>\n\nMYPY_VERSION = "mypy==1.10.0"',
 )
 s.replace(
     "noxfile.py", r'"blacken",', '\g<0>\n    "mypy",',

